File: backend-express-legacy/src/modules/placements/evaluation.repository.ts
import type { Pool, PoolClient, QueryResultRow } from "pg";

import type {
  InternshipEvaluationDto,
  InternshipEvaluationRole,
  InternshipEvaluationSubmit,
  InternshipEvaluationView,
} from "./evaluation.types.js";
import type { PlacementStatus } from "./placement.types.js";

export type InternshipEvaluationDatabase = Pick<Pool, "query" | "connect">;
type Queryable = Pick<Pool, "query"> | PoolClient;

type PlacementContextRow = QueryResultRow & {
  placement_id: string;
  application_id: string;
  placement_status: PlacementStatus;
  expected_start_date: string;
  actual_start_date: string | null;
  completed_date: string | null;
  student_profile_id: string;
  student_user_id: string;
  student_name: string;
  company_id: string;
  company_name: string;
  job_title: string;
};

type EvaluationRow = QueryResultRow & {
  id: string;
  placement_id: string;
  respondent_role: InternshipEvaluationRole;
  submitted_by_user_id: string;
  submitted_by_name: string;
  command_id: string;
  work_quality_rating: number;
  collaboration_rating: number;
  professionalism_rating: number;
  overall_rating: number;
  recommendation: boolean;
  strengths: string;
  improvements: string | null;
  created_at: Date;
};

const contextSelect = `
  SELECT ip.id AS placement_id, ip.application_id, ip.status AS placement_status,
         ip.expected_start_date::text, ip.actual_start_date::text, ip.completed_date::text,
         sp.id AS student_profile_id, sp.user_id AS student_user_id, sp.full_name AS student_name,
         c.id AS company_id, c.name AS company_name, j.title AS job_title
  FROM internship_placements ip
  JOIN applications a ON a.id = ip.application_id
  JOIN student_profiles sp ON sp.id = a.student_profile_id
  JOIN job_posts j ON j.id = a.job_post_id
  JOIN companies c ON c.id = j.company_id`;

const evaluationSelect = `
  SELECT ie.*,
         COALESCE(sp.full_name, cu.full_name, u.email) AS submitted_by_name
  FROM internship_evaluations ie
  JOIN users u ON u.id = ie.submitted_by_user_id
  LEFT JOIN student_profiles sp ON sp.user_id = u.id
  LEFT JOIN company_users cu ON cu.user_id = u.id`;

function mapEvaluation(row: EvaluationRow): InternshipEvaluationDto {
  return {
    id: row.id,
    respondentRole: row.respondent_role,
    workQualityRating: row.work_quality_rating,
    collaborationRating: row.collaboration_rating,
    professionalismRating: row.professionalism_rating,
    overallRating: row.overall_rating,
    recommendation: row.recommendation,
    strengths: row.strengths,
    improvements: row.improvements,
    submittedBy: { id: row.submitted_by_user_id, name: row.submitted_by_name },
    submittedAt: row.created_at.toISOString(),
  };
}

export class InternshipEvaluationRepository {
  constructor(private readonly database: InternshipEvaluationDatabase) {}

  async withTransaction<T>(callback: (client: PoolClient) => Promise<T>) {
    const client = await this.database.connect();
    try {
      await client.query("BEGIN");
      const result = await callback(client);
      await client.query("COMMIT");
      return result;
    } catch (error) {
      await client.query("ROLLBACK");
      throw error;
    } finally {
      client.release();
    }
  }

  async findContextByApplication(applicationId: string, client: Queryable = this.database, lock = false) {
    const result = await client.query<PlacementContextRow>(
      `${contextSelect} WHERE ip.application_id = $1${lock ? " FOR UPDATE OF ip" : ""}`,
      [applicationId],
    );
    return result.rows[0] ?? null;
  }

  async listForPlacement(placementId: string, client: Queryable = this.database) {
    const result = await client.query<EvaluationRow>(
      `${evaluationSelect} WHERE ie.placement_id = $1 ORDER BY ie.created_at, ie.id`,
      [placementId],
    );
    return result.rows.map(mapEvaluation);
  }

  async findByCommand(placementId: string, commandId: string, client: Queryable = this.database) {
    const result = await client.query<EvaluationRow>(
      `${evaluationSelect} WHERE ie.placement_id = $1 AND ie.command_id = $2`,
      [placementId, commandId],
    );
    return result.rows[0] ? mapEvaluation(result.rows[0]) : null;
  }

  async insert(client: PoolClient, input: InternshipEvaluationSubmit, context: PlacementContextRow) {
    await client.query(
      `INSERT INTO internship_evaluations
       (placement_id, respondent_role, submitted_by_user_id, command_id,
        work_quality_rating, collaboration_rating, professionalism_rating, overall_rating,
        recommendation, strengths, improvements)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)`,
      [
        context.placement_id,
        input.role,
        input.actor.userId,
        input.commandId,
        input.input.workQualityRating,
        input.input.collaborationRating,
        input.input.professionalismRating,
        input.input.overallRating,
        input.input.recommendation,
        input.input.strengths,
        input.input.improvements ?? null,
      ],
    );
    const evaluation = await this.findByCommand(context.placement_id, input.commandId, client);
    if (!evaluation) throw new Error("Không thể đọc lại phiếu đánh giá vừa tạo.");
    await client.query(
      `INSERT INTO audit_logs
       (actor_user_id, action, target_type, target_id, metadata, ip_address, user_agent)
       VALUES ($1, $2, 'INTERNSHIP_EVALUATION', $3,
               jsonb_build_object('placementId', $4::text, 'applicationId', $5::text,
                                  'respondentRole', $6::text, 'commandId', $7::text), $8, $9)`,
      [
        input.actor.userId,
        input.role === "COMPANY"
          ? "COMPANY_INTERNSHIP_EVALUATION_SUBMITTED"
          : "STUDENT_INTERNSHIP_EVALUATION_SUBMITTED",
        evaluation.id,
        context.placement_id,
        context.application_id,
        input.role,
        input.commandId,
        input.request.ipAddress,
        input.request.userAgent,
      ],
    );
    await this.createNotifications(client, input, context, evaluation.id);
    return evaluation;
  }

  private async createNotifications(
    client: PoolClient,
    input: InternshipEvaluationSubmit,
    context: PlacementContextRow,
    evaluationId: string,
  ) {
    if (input.role === "COMPANY") {
      await client.query(
        `INSERT INTO notifications
         (recipient_user_id, type, title, body, resource_type, resource_id, deep_link, dedupe_key, payload)
         VALUES ($1, 'COMPANY_INTERNSHIP_EVALUATION_SUBMITTED', 'Doanh nghiệp đã gửi đánh giá kỳ thực tập',
                 $2, 'INTERNSHIP_EVALUATION', $3::uuid, '/applications/' || $4::text,
                 'evaluation:' || $3::text || ':student',
                 jsonb_build_object('placementId', $5::text, 'applicationId', $4::text,
                                    'respondentRole', 'COMPANY'))
         ON CONFLICT (dedupe_key) DO NOTHING`,
        [
          context.student_user_id,
          `${context.company_name} đã gửi đánh giá cho vị trí “${context.job_title}”.`,
          evaluationId,
          context.application_id,
          context.placement_id,
        ],
      );
    } else {
      await client.query(
        `INSERT INTO notifications
         (recipient_user_id, type, title, body, resource_type, resource_id, deep_link, dedupe_key, payload)
         SELECT u.id, 'STUDENT_INTERNSHIP_EVALUATION_SUBMITTED',
                'Sinh viên đã gửi phản hồi kỳ thực tập', $2,
                'INTERNSHIP_EVALUATION', $3::uuid, '/company/candidates/' || $4::text,
                'evaluation:' || $3::text || ':company:' || u.id::text,
                jsonb_build_object('placementId', $5::text, 'applicationId', $4::text,
                                   'respondentRole', 'STUDENT', 'contentVisible', false)
         FROM company_users cu JOIN users u ON u.id = cu.user_id
         WHERE cu.company_id = $1 AND u.status = 'ACTIVE'
         ON CONFLICT (dedupe_key) DO NOTHING`,
        [
          context.company_id,
          `${context.student_name} đã gửi phản hồi riêng cho UIT về vị trí “${context.job_title}”.`,
          evaluationId,
          context.application_id,
          context.placement_id,
        ],
      );
    }
    await client.query(
      `INSERT INTO notifications
       (recipient_user_id, type, title, body, resource_type, resource_id, deep_link, dedupe_key, payload)
       SELECT u.id, $1, $2, $3, 'INTERNSHIP_EVALUATION', $4::uuid,
              '/uit/placements/' || $5::text,
              'evaluation:' || $4::text || ':uit:' || u.id::text,
              jsonb_build_object('placementId', $5::text, 'applicationId', $6::text,
                                 'respondentRole', $7::text)
       FROM uit_staff us JOIN users u ON u.id = us.user_id
       WHERE u.status = 'ACTIVE'
       ON CONFLICT (dedupe_key) DO NOTHING`,
      [
        input.role === "COMPANY"
          ? "COMPANY_INTERNSHIP_EVALUATION_SUBMITTED"
          : "STUDENT_INTERNSHIP_EVALUATION_SUBMITTED",
        input.role === "COMPANY" ? "Doanh nghiệp đã gửi đánh giá thực tập" : "Sinh viên đã gửi phản hồi thực tập",
        `${input.role === "COMPANY" ? context.company_name : context.student_name} đã hoàn tất phiếu cho vị trí “${context.job_title}”.`,
        evaluationId,
        context.placement_id,
        context.application_id,
        input.role,
      ],
    );
  }

  async viewFor(
    context: PlacementContextRow,
    viewerRole: InternshipEvaluationRole,
    client: Queryable = this.database,
  ): Promise<InternshipEvaluationView> {
    const evaluations = await this.listForPlacement(context.placement_id, client);
    const companyEvaluation = evaluations.find((item) => item.respondentRole === "COMPANY") ?? null;
    const rawStudentEvaluation = evaluations.find((item) => item.respondentRole === "STUDENT") ?? null;
    const ownEvaluation = viewerRole === "COMPANY" ? companyEvaluation : rawStudentEvaluation;
    return {
      placement: {
        id: context.placement_id,
        applicationId: context.application_id,
        status: context.placement_status,
        expectedStartDate: context.expected_start_date,
        actualStartDate: context.actual_start_date,
        completedDate: context.completed_date,
      },
      companyEvaluation,
      studentEvaluation: viewerRole === "COMPANY" ? null : rawStudentEvaluation,
      studentEvaluationSubmitted: Boolean(rawStudentEvaluation),
      canSubmit: context.placement_status === "COMPLETED" && !ownEvaluation,
    };
  }
}

File: backend/src/main/java/vn/edu/uit/careerhub/placements/EvaluationRepository.java
package vn.edu.uit.careerhub.placements;

import java.sql.Date;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

import org.springframework.jdbc.core.simple.JdbcClient;
import org.springframework.stereotype.Repository;

import vn.edu.uit.careerhub.auth.RequestMetadata;
import vn.edu.uit.careerhub.common.JsonSupport;
import vn.edu.uit.careerhub.placements.PlacementModels.Context;
import vn.edu.uit.careerhub.placements.PlacementModels.EvaluationRole;
import vn.edu.uit.careerhub.placements.PlacementModels.Status;
import vn.edu.uit.careerhub.placements.PlacementRequests.Evaluation;

@Repository
public class EvaluationRepository {
    private static final String CONTEXT="""
      SELECT ip.id placement_id,ip.application_id,ip.status,ip.expected_start_date,ip.actual_start_date,ip.completed_date,
        sp.id student_profile_id,sp.user_id student_user_id,sp.full_name student_name,c.id company_id,c.name company_name,j.title job_title
      FROM internship_placements ip JOIN applications a ON a.id=ip.application_id JOIN student_profiles sp ON sp.id=a.student_profile_id
      JOIN job_posts j ON j.id=a.job_post_id JOIN companies c ON c.id=j.company_id
      """;
    private static final String EVALUATION="""
      SELECT (jsonb_build_object('id',e.id,'respondentRole',e.respondent_role,'workQualityRating',e.work_quality_rating,
        'collaborationRating',e.collaboration_rating,'professionalismRating',e.professionalism_rating,'overallRating',e.overall_rating,
        'recommendation',e.recommendation,'strengths',e.strengths,'improvements',e.improvements,
        'submittedBy',jsonb_build_object('id',e.submitted_by_user_id,'name',COALESCE(sp.full_name,cu.full_name,u.email)),'submittedAt',e.created_at))::text data
      FROM internship_evaluations e JOIN users u ON u.id=e.submitted_by_user_id LEFT JOIN student_profiles sp ON sp.user_id=u.id LEFT JOIN company_users cu ON cu.user_id=u.id
      """;
    private final JdbcClient database;private final JsonSupport json;
    public EvaluationRepository(JdbcClient database,JsonSupport json){this.database=database;this.json=json;}
    public Optional<Context> context(UUID applicationId,boolean lock){return database.sql(CONTEXT+" WHERE ip.application_id=:id"+(lock?" FOR UPDATE OF ip":""))
            .param("id",applicationId).query((r,n)->new Context(r.getObject("placement_id",UUID.class),r.getObject("application_id",UUID.class),Status.valueOf(r.getString("status")),
                    r.getDate("expected_start_date").toLocalDate(),r.getDate("actual_start_date")==null?null:r.getDate("actual_start_date").toLocalDate(),
                    r.getDate("completed_date")==null?null:r.getDate("completed_date").toLocalDate(),r.getObject("student_profile_id",UUID.class),r.getObject("student_user_id",UUID.class),
                    r.getString("student_name"),r.getObject("company_id",UUID.class),r.getString("company_name"),r.getString("job_title"))).optional();}
    public List<Map<String,Object>> list(UUID placementId){return database.sql(EVALUATION+" WHERE e.placement_id=:id ORDER BY e.created_at,e.id").param("id",placementId)
            .query(String.class).list().stream().map(json::object).toList();}
    public Optional<Map<String,Object>> byCommand(UUID placementId,UUID commandId){return database.sql(EVALUATION+" WHERE e.placement_id=:id AND e.command_id=:commandId")
            .param("id",placementId).param("commandId",commandId).query(String.class).optional().map(json::object);}
    public Map<String,Object> view(Context context,EvaluationRole viewer){List<Map<String,Object>> evaluations=list(context.placementId());
        Map<String,Object> company=evaluations.stream().filter(e->"COMPANY".equals(e.get("respondentRole"))).findFirst().orElse(null);
        Map<String,Object> student=evaluations.stream().filter(e->"STUDENT".equals(e.get("respondentRole"))).findFirst().orElse(null);
        Map<String,Object> placement=new LinkedHashMap<>();placement.put("id",context.placementId());placement.put("applicationId",context.applicationId());placement.put("status",context.status().name());
        placement.put("expectedStartDate",context.expectedStartDate().toString());placement.put("actualStartDate",context.actualStartDate()==null?null:context.actualStartDate().toString());placement.put("completedDate",context.completedDate()==null?null:context.completedDate().toString());
        Map<String,Object> result=new LinkedHashMap<>();result.put("placement",placement);result.put("companyEvaluation",company);result.put("studentEvaluation",viewer==EvaluationRole.COMPANY?null:student);
        result.put("studentEvaluationSubmitted",student!=null);Map<String,Object> own=viewer==EvaluationRole.COMPANY?company:student;result.put("canSubmit",context.status()==Status.COMPLETED&&own==null);return result;}
    public UUID insert(Context c,EvaluationRole role,UUID actorId,UUID commandId,Evaluation input,RequestMetadata request){
        UUID id=database.sql("""
          INSERT INTO internship_evaluations(placement_id,respondent_role,submitted_by_user_id,command_id,work_quality_rating,collaboration_rating,
            professionalism_rating,overall_rating,recommendation,strengths,improvements)
          VALUES(:placementId,:role,:actorId,:commandId,:work,:collaboration,:professionalism,:overall,:recommendation,:strengths,:improvements) RETURNING id
          """).param("placementId",c.placementId()).param("role",role.name()).param("actorId",actorId).param("commandId",commandId)
                .param("work",input.workQualityRating()).param("collaboration",input.collaborationRating()).param("professionalism",input.professionalismRating())
                .param("overall",input.overallRating()).param("recommendation",input.recommendation()).param("strengths",input.strengths().strip())
                .param("improvements",input.improvements()==null?null:input.improvements().strip()).query(UUID.class).single();
        database.sql("""
          INSERT INTO audit_logs(actor_user_id,action,target_type,target_id,metadata,ip_address,user_agent)
          VALUES(:actorId,:action,'INTERNSHIP_EVALUATION',:id,jsonb_build_object('placementId',CAST(:placementId AS text),'applicationId',CAST(:applicationId AS text),
            'respondentRole',:role,'commandId',CAST(:commandId AS text)),CAST(:ip AS inet),:agent)
          """).param("actorId",actorId).param("action",role==EvaluationRole.COMPANY?"COMPANY_INTERNSHIP_EVALUATION_SUBMITTED":"STUDENT_INTERNSHIP_EVALUATION_SUBMITTED")
                .param("id",id).param("placementId",c.placementId()).param("applicationId",c.applicationId()).param("role",role.name()).param("commandId",commandId)
                .param("ip",request.ipAddress()).param("agent",request.userAgent()).update();notify(c,role,id);return id;
    }
    private void notify(Context c,EvaluationRole role,UUID evaluationId){
        if(role==EvaluationRole.COMPANY)database.sql("""
          INSERT INTO notifications(recipient_user_id,type,title,body,resource_type,resource_id,deep_link,dedupe_key,payload)
          VALUES(:userId,'COMPANY_INTERNSHIP_EVALUATION_SUBMITTED','Doanh nghiệp đã gửi đánh giá kỳ thực tập',:body,'INTERNSHIP_EVALUATION',:evaluationId,
            '/applications/'||CAST(:applicationId AS text),'evaluation:'||CAST(:evaluationId AS text)||':student',jsonb_build_object('placementId',CAST(:placementId AS text),
            'applicationId',CAST(:applicationId AS text),'respondentRole','COMPANY')) ON CONFLICT(dedupe_key) DO NOTHING
          """).param("userId",c.studentUserId()).param("body",c.companyName()+" đã gửi đánh giá cho vị trí “"+c.jobTitle()+"”.").param("evaluationId",evaluationId)
                .param("applicationId",c.applicationId()).param("placementId",c.placementId()).update();
        else database.sql("""
          INSERT INTO notifications(recipient_user_id,type,title,body,resource_type,resource_id,deep_link,dedupe_key,payload)
          SELECT u.id,'STUDENT_INTERNSHIP_EVALUATION_SUBMITTED','Sinh viên đã gửi phản hồi kỳ thực tập',:body,'INTERNSHIP_EVALUATION',:evaluationId,
            '/company/candidates/'||CAST(:applicationId AS text),'evaluation:'||CAST(:evaluationId AS text)||':company:'||CAST(u.id AS text),
            jsonb_build_object('placementId',CAST(:placementId AS text),'applicationId',CAST(:applicationId AS text),'respondentRole','STUDENT','contentVisible',false)
          FROM company_users cu JOIN users u ON u.id=cu.user_id WHERE cu.company_id=:companyId AND u.status='ACTIVE' ON CONFLICT(dedupe_key) DO NOTHING
          """).param("body",c.studentName()+" đã gửi phản hồi riêng cho UIT về vị trí “"+c.jobTitle()+"”.").param("evaluationId",evaluationId)
                .param("applicationId",c.applicationId()).param("placementId",c.placementId()).param("companyId",c.companyId()).update();
        database.sql("""
          INSERT INTO notifications(recipient_user_id,type,title,body,resource_type,resource_id,deep_link,dedupe_key,payload)
          SELECT u.id,:type,:title,:body,'INTERNSHIP_EVALUATION',:evaluationId,'/uit/placements/'||CAST(:placementId AS text),
            'evaluation:'||CAST(:evaluationId AS text)||':uit:'||CAST(u.id AS text),jsonb_build_object('placementId',CAST(:placementId AS text),
            'applicationId',CAST(:applicationId AS text),'respondentRole',:role) FROM uit_staff s JOIN users u ON u.id=s.user_id WHERE u.status='ACTIVE'
          ON CONFLICT(dedupe_key) DO NOTHING
          """).param("type",role==EvaluationRole.COMPANY?"COMPANY_INTERNSHIP_EVALUATION_SUBMITTED":"STUDENT_INTERNSHIP_EVALUATION_SUBMITTED")
                .param("title",role==EvaluationRole.COMPANY?"Doanh nghiệp đã gửi đánh giá thực tập":"Sinh viên đã gửi phản hồi thực tập")
                .param("body",(role==EvaluationRole.COMPANY?c.companyName():c.studentName())+" đã hoàn tất phiếu cho vị trí “"+c.jobTitle()+"”.")
                .param("evaluationId",evaluationId).param("placementId",c.placementId()).param("applicationId",c.applicationId()).param("role",role.name()).update();
    }
}
